protocol/messenger_communities.go

```diff
@@ -3730,6 +3730,741 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+					}
+
+					importProgress.AddTaskError(discord.ChannelsCreationTask, discord.Error(errmsg))
+					importProgress.StopTask(discord.ChannelsCreationTask)
+					progressUpdates <- importProgress
+					cancel <- request.DiscordChannelID
+					return
+				}
+
+				community = changes.Community
+				for chatID, chat := range changes.ChatsAdded {
+					newChat = CreateCommunityChat(request.CommunityID.String(), chatID, chat, m.getTimesource())
+				}
+
+				progressValue = calculateProgress(3, 3, (float32(3) / 3))
+
+				importProgress.UpdateTaskProgress(discord.ChannelsCreationTask, progressValue)
+				progressUpdates <- importProgress
+			}
+
+			if m.DiscordImportChannelMarkedAsCancelled(request.DiscordChannelID) {
+				importProgress.StopTask(discord.ImportMessagesTask)
+				progressUpdates <- importProgress
+				cancel <- request.DiscordChannelID
+				return
+			}
+
+			messagesToSave, pinMessagesToSave, authorProfilesToSave, messageAttachmentsToDownload :=
+				m.processDiscordMessages(channel, newChat, importProgress, progressUpdates, request.From, community)
+
+			var discordMessages []*protobuf.DiscordMessage
+			for _, msg := range messagesToSave {
+				if msg.ChatMessage.ContentType == protobuf.ChatMessage_DISCORD_MESSAGE {
+					discordMessages = append(discordMessages, msg.GetDiscordMessage())
+				}
+			}
+
+			// We save these messages in chunks, so we don't block the database
+			// for a longer period of time
+			discordMessageChunks := chunkSlice(discordMessages, maxChunkSizeMessages)
+			chunksCount := len(discordMessageChunks)
+
+			for ii, msgs := range discordMessageChunks {
+				m.communitiesManager.LogStdout(fmt.Sprintf("saving %d/%d chunk with %d discord messages", ii+1, chunksCount, len(msgs)))
+				err := m.persistence.SaveDiscordMessages(msgs)
+				if err != nil {
+					m.cleanUpImportChannel(request.CommunityID.String(), newChat.ID)
+					importProgress.AddTaskError(discord.ImportMessagesTask, discord.Error(err.Error()))
+					importProgress.StopTask(discord.ImportMessagesTask)
+					progressUpdates <- importProgress
+					cancel <- request.DiscordChannelID
+
+					return
+				}
+
+				if m.DiscordImportChannelMarkedAsCancelled(request.DiscordChannelID) {
+					importProgress.StopTask(discord.ImportMessagesTask)
+					progressUpdates <- importProgress
+					cancel <- request.DiscordChannelID
+					return
+				}
+
+				// We're multiplying `chunksCount` by `0.25` so we leave 25% for additional save operations
+				// 0.5 are the previous 50% of progress
+				currentCount := ii + 1
+				progressValue := calculateProgress(i+1, totalImportChunkCount, 0.5+(float32(currentCount)/float32(chunksCount))*0.25)
+				importProgress.UpdateTaskProgress(discord.ImportMessagesTask, progressValue)
+				progressUpdates <- importProgress
+
+				// We slow down the saving of message chunks to keep the database responsive
+				if currentCount < chunksCount {
+					time.Sleep(2 * time.Second)
```

Comment:
This is different. The speedup/slowdown was implemented for importing archives, not for importing Discord communities/channels.

The same approach could be applied here as well, but it was never a priority because importing a community is a one-time action. I would stick with the naive `time.Sleep` for now, unless we decide to undertake a significant refactor of the complex `RequestImportDiscordCommunity`.